### ISSUE TITLE
Add option to emulate full :recompile-dependents

### ISSUE THREAD

**Issue by thheller** [thheller/shadow-cljs#349]
shadow-cljs only recompiles direct dependents of namespaces and stops there. Some frameworks/templates seem to rely on the figwheel/CLJS behaviour that reloads all other transitive dependents as well.

Reloading all is only required if namespaces hold global state of other namespaces so it is rarely useful/required. Should still probably have an option to enable this behavior if desired.

**Comment by peterhudec** [comment]
How about a case when I have a function or a macro (no global state) in a transitive dependency?

Let's say that I have a library project with this structure:

```
src/
  my_project/
    core.clj # Macros are here
    core.cljs
  my_project.cljs # This contains all the public APIs
  example.cljs # I'm experimenting here
```

When I make a change in any file in `my_project/`, the watch won't notice.

I'm new to Clojure and ClojureScript, but from what I've seen, projects don't seem to have a flat structure.

**Comment by thheller** [comment]
I'm not sure I understand your point?

What do you expect to happen that isn't happening? Your filestructure isn't useful on its own, need the `ns` setup since they dictate what/how reloading is performed.

**Comment by peterhudec** [comment]
I can actually simplify the problem:

I have a `src/main.cljs` file, which requires the `src/my_project.clj` file. So far so good, when I change `src/my_project.clj`, it reloads.

```clj
;; src/main.cljs
(ns main
  (:require [my-project])
```

```clj
;; src/my_project.clj
(ns my-project)
```

But what if `src/my_project.clj` grows too big and needs to be split into multiple namespaces.

```clj
(ns my-project
  (:require [utils])
```

```clj
;; src/utils.clj
(ns utils)
```
 Now, if I change `src/utils.clj`, there will be no reload.

**Comment by thheller** [comment]
As a general rule you should follow JVM classpath conventions and pick unique-enough names for your files. `utils` and `main` are way too generic and should be grouped into a unique directory. Lets assume `my-project.*` as a ns prefix for everything (assuming my-project is unique enough). This is to avoid conflicts with other libraries that may be on the classpath. If two libraries were to use `(ns utils)` only one would be available at runtime. JVM/Java in general settled on using the reverse domain name systems (eg. `com.company.foo`) to ensure some level of unique-ness but this is less common in the CLJ(S) world.

The "short" names sort of allow you to be lazy and skip requiring other namespaces properly. In general all namespaces that have a `:require [my-project.utils :as u]` will be reloaded, those that don't and just use the ns directly won't be.

**Comment by peterhudec** [comment]
I'm afraid I simplified it a bit too much. Off course I have all the namespaces prefixed.

Nevertheless, after some experimenting I have isolated the problem to this:

Updates to Clojure namespaces required by macros are ignored by the watcher. I have forked the [minimal-shadow-cljs-browser](https://github.com/peterhudec/minimal-shadow-cljs-browser/blob/master/src/app/macro_helpers.clj) example, to demonstrate the problem. Here is an outline of it:

Considering this folder structure:
```
src/
  app/
    main.cljs
    macro.clj
    macros.cljs
    macro_helpers.clj
```

This file is required by `main` and changes to it are noticed by the watcher:
```clj
;; src/app/macros.clj
(ns app.macros
  (:require [app.macro-helpers :as helpers]))

(defmacro m [& forms]
  `(do
     (println (str "I'm a macro and I reload"))
     (println ~(helpers/helper "Foo"))
     ~@forms))
```

This file is only required in the above macro (at compile time?) and changes to it are ignored by the watcher:
```clj
;; src/app/macro_helpers.clj
(ns app.macro-helpers)

(defn helper [x]
  (str "I'm a macro helper and I don't reload. "
       [x x x]))
```

**Comment by thheller** [comment]
Ok yeah that is known. CLJ dependencies aren't tracked beyond the initial macro require. It probably should be though. FWIW when developing complex macros that have other dependencies I usually use a CLJ REPL to do the development and reloading.

**Comment by thheller** [comment]
Macro reloading is fixed. Not going to do `:recompile-dependents` since the only reason to do so is allowing reload to work in a bad project architecture.

**Comment by aiba** [comment]
Can we please reconsider implementing `:recompile-dependents`?

I'm working on a react-native project using [React Navigation](https://reactnavigation.org/) and the development workflow for this library assumes that if you redefine a component, then all the dependent react components will be recompiled.  Without `:recompile-dependents`, I have to reload the entire app each time I make any change to a react-navigation component.

This is a popular and useful NPM library, it would be really nice to be able to use its development workflow.

**Comment by thheller** [comment]
Can you point me towards the docs about "assumes that if you redefine a component, then all the dependent react components will be recompiled"? That is a very weird assumption and not how any reloading mechanism I'm aware of works.

Re-render fine but recompile seems questionable.

**Comment by aiba** [comment]
Yeah, that's a fair question.  The assumption isn't documented, I found it through debugging why the [state-preserving development workflow](https://reactnavigation.org/docs/en/state-persistence.html) was working in javascript but not clojurescript, and the only way I finally got it to work was by manually adding all the transitive dependencies among my clojurescript files (to simulate `:recompile-dependents`).  Then it worked.

I'm working off [this template](https://github.com/thheller/reagent-react-native/) and using the [forceUpdate trick](https://github.com/thheller/reagent-react-native/blob/master/src/main/test/app.cljs#L42) to do hot reloads, rather than react-native's "fast refresh".  It's possible this is part of what's violating the library's assumption.

Reading about [react-native's fast refresh](https://aboutreact.com/fast-refresh-in-react-native/), I see that it does not usually recompile dependents, and has a bunch of other differences from the [reagent-react-native](https://github.com/thheller/reagent-react-native/) live-reloading workflow.

So in conclusion, I can't fully explain why react-navigation live reloading doesn't work with shadow-cljs.  I can say that implementing `:recompile-dependents` would solve it.  But there might be a deeper issue that I'm failing to understand.

If I ever get to the bottom of this, I could report back on this thread.  I realize that as it stands, this is not the most compelling request for the feature.

**Comment by thheller** [comment]
If you setup a reproducible example I can maybe take a look.

You can find a writeup about react-native-navigation here:
https://040code.github.io/2019/04/02/react-native-navigation-and-clojurescript

Dunno how much of it applies to react-navigation but the things you have to do will be pretty similar regarding hot-reload.

**Comment by aiba** [comment]
Wow, thank you!  That would be amazing. This mystery has been driving my crazy.

I boiled it down to a minimal example: https://github.com/aiba/cljs-react-navigation

To reproduce:
1. Run the simulator.
1. Navigate to "Tab B" in the UI.
1. Edit `src/myapp/nav/sub2.cljs` and make changes to the `TabB` component.  You'll see that upon saving the file, shadow-cljs recompiles `myapp.nav.sub2` and its dependent `myapp.nav.sub1`, BUT the react-navigation state is somehow lost so that you are no longer looking at `TabB` in the UI.
1. Note that if you uncomment [line 7 of nav/core](https://github.com/aiba/cljs-react-navigation/blob/master/src/myapp/nav/core.cljs#L7), manually establishing the transitive dependency, then repeating the previous step does not lose the navigation state when editing the `TabB` component.

**Comment by thheller** [comment]
I don't have a working RN setup right now but looking at the code it looks like an regular render short circuit. Meaning that react sees that `[sub1/Root]` did not change and stops rendering.

Can you try `[sub1/Root {:x (js/Date.now)}]` or some kind of other data causing it continue rendering? I'd expect it to work normally after that?

**Comment by aiba** [comment]
In both cases (with the `{:x (js/Date.now)}` and without), the full component tree is re-rendered.  (I confirmed this console.log()s.)  Presumably that's becuase of the `.forceUpdate` in `main.cljs`.  So unfortunately, adding the Date.now doesn't fix it.

My new theory is that the problem has to do with the way the `hx.react/defnc` macro statically compiles down the hiccup template.  Perhaps if that macro is reaching deep in the component tree at compile-time, then it would make sense that the reload isn't working without recompiling the higher-up components.

**Comment by thheller** [comment]
I give up.

I added `:devtools {:reload-strategy :full}` which should fix these kinds of issues without resorting to a full `:recompile-dependents`. Try `2.8.85`. It'll reload all sources as if `:recompile-dependents` was enabled without actually compiling them. Most likely that is enough.

**Comment by Lokeh** [comment]
Just commenting here that, while I am not clear on the cause of the hot loading behavior when using react navigation, hx (and helix) do nothing in terms of directly manipulating or memoizing the component tree.

That being said, the `:reload-strategy :full` is a fantastic option that allows Fast Refresh to be used without much fanfare with helix. It's unfortunate that React's strategy of memoizing all the things requires us to reload all of those namespaces, but gaining the ability to maintain local React component state when reloading seems like a tradeoff many would make for that. Thank you for accommodating it as a part of shadow-cljs!

**Comment by aiba** [comment]
@thheller this change looks great but it seems to only apply to `:target :browser`.  `shadow.cljs.devtools.client.react-native` has a different way of reloading dependencies for react-native.

Would it be possible to have a `:reload-strategy :full` that works with `:target :react-native`?

**Comment by thheller** [comment]
@aiba should be working with react-native in `2.8.87`.

**Comment by aiba** [comment]
I hate to be so annoying about this issue, and I appreciate so much the addition of the full reload strategy.  But I think commit b81ea1a1 appears to have broken all react-native code reloading.  I think it could be because the `module-format` for react-native is `"goog"`, and after this commit, `shadow.cljs.devtools.client.env/prefilter-sources` line 298 is now excluding all sources from being reloaded.

**Comment by thheller** [comment]
hmm yeah you are right.

I guess I need to get a working react-native setup again to test this stuff. The emulator stuff I had setup last time won't start anymore and I don't have the patience to figure out why.

Your diagnosis is incorrect though, its the second check that fails.

**Comment by aiba** [comment]
Everything working great in 2.8.88.  Thank you!  :star: :tada:

**Comment by yannvanhalewyn** [comment]
Wanted to leave a note of appreciation for `:reload-strategy :full`. I was running into the same issues as @aiba with React Navigation (native) and this fixed it. It shouldn't be necessary, it seems like React Navigation makes some wrong assumptions in their rendering lifecycle, but I'm so glad there is a solution. 

The Shadow keeps us in the light, so thank you so much @thheller! 🎉

**Comment by aiba** [comment]
FWIW, to get hot reloading working with react-navigation, I now use [lilactown/helix](https://github.com/lilactown/helix) with [fast refresh](https://github.com/lilactown/helix/blob/master/docs/experiments.md#fast-refresh) and it works seamlessly.

Also still using shadow-cljs for react-native development and it's working great also.  Thank you @thheller!
